// @funeste38/qflush/dist/diagnostics/collect.js
// ROME-TAG: 0xD1AG — diagnostics collector
import * as fs from 'node:fs';
import * as path from 'node:path';
import os from 'node:os';
import { resolvePaths, SERVICE_MAP } from '../utils/paths.js';
import { detectModules } from '../utils/detect.js';
export async function collectDiagnostics(startTime, optsServices) {
    const errors = [];
    const now = Date.now();
    let detectedModules = [];
    let resolvedPaths = {};
    let spyderPort = undefined;
    let cortexCount = null;
    try {
        // detectModules returns a Record in this codebase; coerce to string[] keys if object
        const detected = await detectModules();
        if (Array.isArray(detected))
            detectedModules = detected;
        else if (detected && typeof detected === 'object')
            detectedModules = Object.keys(detected);
    }
    catch (e) {
        errors.push(`detectModules failed: ${String(e)}`);
    }
    try {
        resolvedPaths = resolvePaths(detectedModules);
    }
    catch (e) {
        errors.push(`resolvePaths failed: ${String(e)}`);
    }
    try {
        const p = path.join(process.cwd(), '.qflush', 'spyder.config.json');
        if (fs.existsSync(p)) {
            const raw = fs.readFileSync(p, 'utf8') || '{}';
            const cfg = JSON.parse(raw);
            if (cfg?.adminPort)
                spyderPort = Number(cfg.adminPort);
        }
    }
    catch (e) {
        errors.push(`spyder.config.json parse error: ${String(e)}`);
    }
    try {
        const routes = path.join(process.cwd(), '.qflush', 'cortex.routes.json');
        if (fs.existsSync(routes)) {
            const raw = fs.readFileSync(routes, 'utf8') || '{}';
            const parsed = JSON.parse(raw);
            const items = parsed.routes ?? parsed.cortexActions ?? null;
            if (Array.isArray(items))
                cortexCount = items.length;
            else if (items && typeof items === 'object')
                cortexCount = Object.keys(items).length;
        }
    }
    catch (e) {
        errors.push(`cortex.routes.json parse error: ${String(e)}`);
    }
    const system = {
        platform: os.platform(),
        arch: os.arch(),
        cpus: os.cpus()?.length ?? 0,
        memory: {
            total: os.totalmem(),
            free: os.freemem(),
        },
        node: process.version,
        pid: process.pid,
        cwd: process.cwd(),
    };
    const services = optsServices?.length ? optsServices : Object.keys(SERVICE_MAP);
    const diag = {
        timestamp: new Date().toISOString(),
        system,
        qflush: {
            detectedModules,
            resolvedPaths,
            services,
            spyderPort,
            cortexRoutes: cortexCount,
        },
        errors,
        durationMs: now - startTime,
    };
    return diag;
}
