// @funeste38/beam/funebeam.ts
/**
 * FUNEBEAM — Moteur universel Funesterie
 * Supporte deux vocabulaires :
 *
 * TECHNIQUE :
 *  - beam
 *  - drill
 *  - bomb
 *  - fusion
 *  - ultra
 *
 * FUNESTERIE / DBZ :
 *  - kamehameha  → beam
 *  - makankosappo → drill
 *  - genkidama   → bomb
 *  - gogeta      → fusion
 *  - ssj, god, ultra, limitbreaker → ultra
 *
 * Le moteur traduit automatiquement les alias.
 */

export type BeamType = "beam" | "drill" | "bomb" | "fusion" | "ultra";

export interface BeamJob {
  type: string; // accepté : noms techniques ou noms DBZ
  source: string;
  target: string;
  payload?: unknown;
  meta?: Record<string, unknown>;
}

export interface BeamResult {
  ok: boolean;
  type: BeamType;
  logs?: string[];
  data?: unknown;
}

/**
 * Table de conversion Funesterie → pipeline technique
 */
export const FUNEBEAM_ALIASES: Record<string, BeamType> = {
  // BEAM
  beam: "beam",
  kame: "beam",
  kameha: "beam",
  kamehameha: "beam",

  // DRILL
  drill: "drill",
  mkp: "drill",
  makanko: "drill",
  makankosappo: "drill",
  specialbeam: "drill",

  // BOMB
  bomb: "bomb",
  spiritbomb: "bomb",
  genki: "bomb",
  genkidama: "bomb",

  // FUSION
  fusion: "fusion",
  gogeta: "fusion",
  vegito: "fusion",
  fusiondance: "fusion",

  // ULTRA
  ultra: "ultra",
  ssj: "ultra",
  ssj2: "ultra",
  ssjblue: "ultra",
  god: "ultra",
  superfunesterie: "ultra",
  limitbreaker: "ultra",
};

export interface BeamHandlers {
  beam: (job: BeamJob, ctx: BeamContext) => Promise<BeamResult>;
  drill: (job: BeamJob, ctx: BeamContext) => Promise<BeamResult>;
  bomb: (job: BeamJob, ctx: BeamContext) => Promise<BeamResult>;
  fusion: (job: BeamJob, ctx: BeamContext) => Promise<BeamResult>;
  ultra: (job: BeamJob, ctx: BeamContext) => Promise<BeamResult>;
}

export interface BeamContext {
  handlers: BeamHandlers;
  services?: Record<string, unknown>;
  log?: (msg: string) => void;
}

function normalizeBeamType(type: string): BeamType {
  const key = type.toLowerCase();
  const mapped = FUNEBEAM_ALIASES[key];
  if (!mapped) {
    throw new Error(`Unknown beam type or alias: "${type}"`);
  }
  return mapped;
}

export async function dispatchBeam(job: BeamJob, ctx: BeamContext): Promise<BeamResult> {
  const logs: string[] = [];
  const log = (m: string) => {
    logs.push(m);
    ctx.log?.(m);
  };

  let resolvedType: BeamType;

  try {
    resolvedType = normalizeBeamType(job.type);
  } catch (e: any) {
    return { ok: false, type: "beam", logs: [`❌ ${e.message}`] };
  }

  const handler = ctx.handlers[resolvedType];
  if (!handler) {
    return { ok: false, type: resolvedType, logs: [`❌ No handler for type "${resolvedType}"`] };
  }

  log(`🚀 FUNEBEAM Dispatch: alias "${job.type}" → pipeline "${resolvedType}"`);

  try {
    const result = await handler(job, ctx);
    result.type = resolvedType;
    result.logs = [...(result.logs || []), ...logs];
    return result;
  } catch (err: any) {
    return {
      ok: false,
      type: resolvedType,
      logs: [...logs, `💥 Handler error: ${String(err?.message || err)}`],
    };
  }
}
