// @funeste38/qflush/dist/tests/strict-runner.js
// ROME-TAG: 0xTEST01 — strict-runner
import { execSync } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';
function runOneTest(file) {
    const start = Date.now();
    try {
        execSync(`vitest run "${file}"`, {
            stdio: 'pipe',
            encoding: 'utf8',
        });
        return {
            file,
            ok: true,
            duration: Date.now() - start,
        };
    }
    catch (err) {
        return {
            file,
            ok: false,
            duration: Date.now() - start,
            error: String(err?.stdout || err?.message || err),
        };
    }
}
export function listTests() {
    const root = path.join(process.cwd(), 'src', 'tests');
    if (!fs.existsSync(root))
        return [];
    const results = [];
    const scan = (dir) => {
        for (const f of fs.readdirSync(dir)) {
            const full = path.join(dir, f);
            if (fs.statSync(full).isDirectory())
                scan(full);
            else if (full.endsWith('.test.ts'))
                results.push(full);
        }
    };
    scan(root);
    return results;
}
export async function runStrictSuite() {
    const tests = listTests();
    if (!tests.length) {
        console.log('No tests found.');
        return;
    }
    const summary = [];
    for (const t of tests) {
        console.log(`🚀 Running: ${path.basename(t)}...`);
        const result = runOneTest(t);
        summary.push(result);
        if (!result.ok) {
            console.log(`❌ Failed: ${path.basename(t)}`);
            console.log(result.error);
        }
        else {
            console.log(`✅ OK: ${path.basename(t)} (${result.duration}ms)`);
        }
    }
    console.log('\n===== SUMMARY =====');
    let pass = 0;
    let fail = 0;
    for (const r of summary) {
        if (r.ok)
            pass++;
        else
            fail++;
    }
    console.log(`Passed: ${pass}`);
    console.log(`Failed: ${fail}`);
    if (fail > 0)
        process.exit(1);
}
