// @funeste38/qflush/dist/utils/refactor.js
import fs from 'node:fs';
import path from 'node:path';
import ts from 'typescript';
function findAllTSFiles() {
    const root = process.cwd();
    const result = [];
    function walk(dir) {
        if (!fs.existsSync(dir))
            return;
        for (const name of fs.readdirSync(dir)) {
            if (name.startsWith('.'))
                continue;
            const full = path.join(dir, name);
            const st = fs.statSync(full);
            if (st.isDirectory())
                walk(full);
            else if (full.endsWith('.ts'))
                result.push(full);
        }
    }
    walk(path.join(root, 'src'));
    const pkgs = path.join(root, 'packages');
    if (fs.existsSync(pkgs))
        walk(pkgs);
    return result;
}
function transformSource(sourceText, fileName) {
    const sf = ts.createSourceFile(fileName, sourceText, ts.ScriptTarget.ESNext, true, ts.ScriptKind.TS);
    function visitor(node) {
        // 1) Replace `if (!expr)` where expr is identifier/property by `if (expr == null)`
        if (ts.isIfStatement(node)) {
            const cond = node.expression;
            if (ts.isPrefixUnaryExpression(cond) && cond.operator === ts.SyntaxKind.ExclamationToken) {
                const operand = cond.operand;
                if (ts.isIdentifier(operand) ||
                    ts.isPropertyAccessExpression(operand) ||
                    ts.isElementAccessExpression(operand)) {
                    const nullLit = ts.factory.createNull();
                    const newCond = ts.factory.createBinaryExpression(operand, ts.SyntaxKind.EqualsEqualsToken, nullLit);
                    return ts.factory.updateIfStatement(node, newCond, ts.visitNode(node.thenStatement, visitor), node.elseStatement ? ts.visitNode(node.elseStatement, visitor) : undefined);
                }
            }
        }
        // 2) Replace empty catch blocks `catch { }` or catch with empty body
        if (ts.isCatchClause(node)) {
            const block = node.block;
            if (block.statements.length === 0) {
                // ensure a variable name exists for the catch
                const varName = node.variableDeclaration && ts.isIdentifier(node.variableDeclaration.name) ? node.variableDeclaration.name : ts.factory.createIdentifier('err');
                const callExpr = ts.factory.createCallExpression(ts.factory.createPropertyAccessExpression(ts.factory.createIdentifier('console'), 'warn'), undefined, [ts.factory.createCallExpression(ts.factory.createIdentifier('String'), undefined, [varName])]);
                const stmt = ts.factory.createExpressionStatement(callExpr);
                const newBlock = ts.factory.createBlock([stmt], true);
                const newDecl = node.variableDeclaration ? node.variableDeclaration : ts.factory.createVariableDeclaration(varName);
                return ts.factory.createCatchClause(newDecl, newBlock);
            }
        }
        return ts.visitEachChild(node, visitor, transformationContext);
    }
    // create a transformation context placeholder to satisfy API typing; will be set below
    let transformationContext;
    const transformer = (context) => {
        transformationContext = context;
        return (rootNode) => ts.visitNode(rootNode, visitor);
    };
    const result = ts.transform(sf, [transformer]);
    const transformedSourceFile = result.transformed[0];
    const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
    const out = printer.printFile(transformedSourceFile);
    result.dispose();
    return out;
}
function fixTextual(source) {
    let s = source;
    // remove `as any` occurrences (naive)
    s = s.replace(/\s+as\s+any/g, '');
    // replace `any | X` with `unknown | X` conservatively
    s = s.replace(/\bany\s*\|/g, 'unknown |');
    // change simple promise check pattern: `if (p && typeof p.then === 'function') p.catch(...)` -> optional chaining
    s = s.replace(/if\s*\(\s*([a-zA-Z0-9_.$]+)\s*&&\s*typeof\s+\1\.then\s*===\s*['"]function['"]\s*\)\s*([a-zA-Z0-9_.$]+)\.catch\(([^)]+)\);/g, (m, p1, p2, args) => {
        // ensure same identifier
        if (p1 === p2) {
            return `if (typeof ${p1}?.then === 'function') { ${p1}.catch(${args}); }`;
        }
        return m;
    });
    // replace empty catches produced as `catch (err) {}` already handled by AST, but keep naive replacement
    s = s.replace(/catch\s*\(\s*\w*\s*\)\s*\{\s*\}/g, `catch (err) { console.warn(String(err)); }`);
    return s;
}
export async function runAutoRefactor() {
    const files = findAllTSFiles();
    for (const file of files) {
        try {
            const src = fs.readFileSync(file, 'utf8');
            const astOut = transformSource(src, file);
            const final = fixTextual(astOut);
            if (final !== src) {
                fs.writeFileSync(file, final, 'utf8');
                console.log('[refactor] updated', file);
            }
        }
        catch (err) {
            // log and continue
            console.warn('[refactor] failed to process', file, String(err));
        }
    }
}
